### PR TITLE
Fix MiniMax-M2.7 on CPU

### PR DESCRIPTION
**Pull request by chunyuan-w**
## Motivation
Fixes MiniMax-M2.7 on CPU

## Modifications
- Make `fused_topk_cpu` fall back to torch native when `correction_bias is not None or scoring_func != "softmax"` since this is unsupported in kernel yet.
- Add `_forward_cpu` to `MiniMaxM2QKRMSNorm` and dispatch to it on CPU.
- Handle uneven TP sharding in `MiniMaxM2RMSNormTP.weight_loader`.

## Accuracy Tests
gsm8k: 0.930
hellaswag: 0.750
```sh
# Server
python3 -m sglang.launch_server --model MiniMaxAI/MiniMax-M2.7 --disable-radix-cache --trust-remote-code --device cpu --disable-overlap-schedule --tp 6 --tool-call-parser minimax-m2 --reasoning-parser minimax-append-think

# Client
# gsm8k
python -m sglang.test.few_shot_gsm8k --num-questions 200 --num-shots 5

# hellaswag
cd benchmark/hellaswag
python3 bench_sglang.py --num-questions 200 --num-shots 20
```

## Speed Tests and Profiling

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #26491147742](https://github.com/sgl-project/sglang/actions/runs/26491147742)
Latest PR Test (Extra): :x: [Run #26491147648](https://github.com/sgl-project/sglang/actions/runs/26491147648)

### CODE REVIEW

**Comment by chunyuan-w** [comment]
@mingfeima could you please review this PR? CI failures are unrelated.

**Comment by mingfeima** [comment]
cc @ZailiWang

File: python/sglang/srt/layers/quantization/fp8.py [hunk 1]
```diff
@@ -1699,6 +1699,11 @@ def apply(
                 moe_runner_config.apply_router_weight_on_input, topk_weights, x
             )
 
+            # fused_topk_torch_native (CPU fallback for models like MiniMax)
+            # returns int64 topk_ids; fused_experts_cpu requires int32.
+            if topk_ids.dtype == torch.int64:
+                topk_ids = topk_ids.to(torch.int32)
+
```
Comment (mingfeima) [inline python/sglang/srt/layers/quantization/fp8.py]: hide this work around inside `fused_experts_cpu`, don't expose it python.
mark a TODO to remove it after topk kernel is provided.

File: python/sglang/srt/layers/logits_processor.py [hunk 2]
```diff
@@ -539,11 +540,11 @@ def _get_pruned_states(
             # Build the index tensors via pinned host memory + non-blocking H2D
             # so the small copy doesn't drain the stream.
             sample_indices = torch.tensor(
-                sample_indices, dtype=torch.int64, pin_memory=True
-            ).to(pruned_states.device, non_blocking=True)
+                sample_indices, dtype=torch.int64, pin_memory=not _is_cpu
```
Comment (mingfeima) [inline python/sglang/srt/layers/logits_processor.py]: use https://github.com/sgl-project/sglang/blob/c317beda99dc723010023e8a6e828cc3145c9ec0/python/sglang/srt/platforms/interface.py#L91

`is_pin_memory_available`

File: python/sglang/srt/layers/logits_processor.py [hunk 3]
```diff
@@ -613,8 +614,8 @@ def _expand_metadata_for_logprobs(
         pruned_lens = torch.tensor(
             logits_metadata.extend_logprob_pruned_lens_cpu,
             dtype=torch.int64,
-            pin_memory=True,
-        ).to(device, non_blocking=True)
+            pin_memory=not _is_cpu,
+        ).to(device, non_blocking=not _is_cpu)
```
Comment (mingfeima) [inline python/sglang/srt/layers/logits_processor.py]: same as above.

File: python/sglang/srt/models/minimax_m2.py [hunk 4]
```diff
@@ -100,6 +102,7 @@
 from sglang.srt.utils.hf_transformers_utils import get_rope_config
 
 logger = logging.getLogger(__name__)
+_is_cpu = is_cpu()
```
Comment (mingfeima) [inline python/sglang/srt/models/minimax_m2.py:106]:
```suggestion
_is_cpu = is_cpu()
_is_amx_available = cpu_has_amx_support()
```

File: python/sglang/srt/models/minimax_m2.py [hunk 5]
```diff
@@ -311,6 +314,20 @@ def weight_loader(
         """Custom weight loader that handles TP sharding."""
         shard_id = self.attn_tp_rank // self.num_head_replicas
         shard_size = param.data.shape[0]
+
+        if _is_cpu:
```
Comment (mingfeima) [inline python/sglang/srt/models/minimax_m2.py]:
```suggestion
        if _is_cpu and _is_amx_available :
```

**Review by mingfeima — CHANGES_REQUESTED** [review]
Some minor places to address. Overall LGTM.

**Comment by mingfeima** [comment]
@chunyuan-w please also open an issue to track TODOs for m2.7 - what needs to be further optimized?

**Comment by mingfeima** [comment]
@ZailiWang update cook book once this one settled.

File: python/sglang/srt/models/minimax_m2.py (hunk 5 above)
Comment (chunyuan-w) [reply python/sglang/srt/models/minimax_m2.py]: Fixed.

File: python/sglang/srt/models/minimax_m2.py (hunk 4 above)
Comment (chunyuan-w) [reply python/sglang/srt/models/minimax_m2.py:106]: Fixed.

File: python/sglang/srt/layers/quantization/fp8.py (hunk 1 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/quantization/fp8.py]: Fixed.

File: python/sglang/srt/layers/logits_processor.py (hunk 2 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/logits_processor.py]: Updated the code to set `pin_memory` to the value of `is_pin_memory_available()`. For `non_blocking`, it can actually be always set to `True`.

File: python/sglang/srt/layers/logits_processor.py (hunk 3 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/logits_processor.py]: This has been fixed on the latest main. After the rebase, this change is not needed.

File: python/sglang/srt/layers/logits_processor.py (hunk 2 above)
Comment (chunyuan-w) [reply python/sglang/srt/layers/logits_processor.py]: I noticed that this has been fixed on the latest main. After the rebase, this change is not needed.

**Comment by chunyuan-w** [comment]
> @chunyuan-w please also open an issue to track TODOs for m2.7 - what needs to be further optimized?

https://github.com/sgl-project/sglang/issues/26439 is created to track the TODOs.

**Review by mingfeima — APPROVED** [review]
